hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/OperationRunner.java

```diff
@@ -151,26 +158,45 @@ public final int getPartitionId() {
     }
 
     /**
-     * Runs operation directly without checking any conditions;
+     * Runs op directly without checking any conditions;
      * node state, partition ownership, timeouts etc.
      * <p>
      * {@link Operation#beforeRun()}, {@link Operation#call()}
      * and {@link Operation#afterRun()} phases are executed sequentially.
      * <p>
      * Operation responses and backups are ignored.
      *
-     * @param operation operation to run
-     * @throws Exception when one of the operation phases fails with an exception
+     * @param op op to run
+     * @throws Exception when one of the op phases fails with an exception
      */
-    public static void runDirect(Operation operation) throws Exception {
+    public static void runDirect(Operation op) throws Exception {
+        try {
+            op.pushThreadContext();
+            op.beforeRun();
+            op.call();
+            op.afterRun();
+        } finally {
+            op.popThreadContext();
+            op.afterRunFinal();
+        }
+    }
+
+    public static void runDirect(Operation op, NodeEngineImpl nodeEngine,
+                                 Set<Operation> asyncOperations) throws Exception {
         try {
-            operation.pushThreadContext();
-            operation.beforeRun();
-            operation.call();
-            operation.afterRun();
+            op.pushThreadContext();
+            op.beforeRun();
+            CallStatus callStatus = op.call();
+            if (callStatus.ordinal() == OFFLOAD_ORDINAL) {
+                op.afterRun();
```

Comment:
is it intended to call `op.afterRun()` twice for offloaded operations?

Reply:
no need that, fixed it.